### PR TITLE
Rewrite policyfile changelog feature

### PR DESCRIPTION
**Pull request by dlukman**
It has less integration with ChefDK classes so it isn't so dependent specific code that might change in the future. Git interaction is simplified using the gem. Improved test coverage. Shows changelog to updated version instead of master. Some minor rubocop style fixes applied to modified code.

### CODE REVIEW

File: lib/knife/changelog/policyfile.rb [hunk 1]
```diff
@@ -1,69 +1,158 @@
...
-      end
+    lock_target = read_policyfile_lock(update_policyfile_lock)
+    target = versions(lock_target['cookbook_locks'], 'target')
 
-      # return a list of supermarket uri for a given cookbook
-      # example: [ 'https://supermarket.chef.io' ]
-      def supermarkets_for(name)
-        [policy.send(:best_source_for, name).uri]
-      end
+    cookbooks = current.deep_merge(target).reject { |_name, data| reject_version_filter(data) }
```
Comment (clems4ever) [inline lib/knife/changelog/policyfile.rb]: OPT: rename into updated_cookbooks

File: lib/knife/changelog/policyfile.rb [hunk 2]
```diff
@@ -1,69 +1,158 @@
 # frozen_string_literal: true
```
Comment (clems4ever) [inline lib/knife/changelog/policyfile.rb]: The code is very clean and contract of methods very clear, this is really great and it can part of counterbalance the following! The following is optional: as discussed together to clarify the contract of the interface, I would have make almost all methods private. Actually only one method needs to be in the public contract of this class. If you did this for testability (what I can understand) I would have spreaded the concerns into several classes that have their own public contracts and which are therefore testable and composable. That way you both solve testability and 1 class = 1 responsibility which will ease the identification of impacts of future fixes (https://en.wikipedia.org/wiki/Single_responsibility_principle)

File: spec/unit/policyfile_spec.rb [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+        }
+        expect(changelog.versions(locks, 'target')).to eq(target_versions)
+      end
+    end
+
+    context 'when type is not current nor target' do
+      it 'raises and exception' do
+        expect { changelog.versions(nil, 'toto') }.to raise_error(RuntimeError)
+      end
+    end
```
Comment (clems4ever) [inline spec/unit/policyfile_spec.rb]: OPT: add test for when locks is empty? (an easy one just for completeness but really not necessary)

File: spec/unit/policyfile_spec.rb [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+        expect(changelog.tag_format(repo)).to eq('')
+      end
+
+      it 'detects type for v-tag' do
+        allow(repo).to receive_message_chain(:tags, :last, :name).and_return('v1.0.0')
+        expect(changelog.tag_format(repo)).to eq('v')
+      end
+    end
+  end
+end
```
Comment (clems4ever) [inline spec/unit/policyfile_spec.rb]: Unfortunately, there is no test for generate_changelog which is the method you definitely need to test, this is your public contract.

File: lib/knife/changelog/policyfile.rb [hunk 5]
```diff
@@ -1,69 +1,158 @@
...
+require 'deep_merge'
+require 'git'
+require 'json'
+require 'rest-client'
 
-class KnifeChangelog
-  class Changelog
-    class Policyfile < Changelog
-      attr_reader :policy, :lock
+class PolicyChangelog
```
[the remark this answers is not in the document]
Comment (clems4ever) [reply lib/knife/changelog/policyfile.rb]: This is actually true. +1

[the remark this answers is not in the document]
Comment (dlukman) [reply lib/knife/changelog/policyfile.rb]: Moved to lib

File: lib/knife/changelog/policyfile.rb [hunk 6]
```diff
@@ -1,69 +1,158 @@
...
+  # Clones a Git repo in a temporary directory and generates a commit
+  # changelog between two version tags
+  #
+  # @param source_url [String] Git repository URL
+  # @param current [String] current cookbook version tag
+  # @param target [String] target cookbook version tag
+  # @return [String] changelog between tags for one cookbook
+  def git_changelog(source_url, current, target)
+    dir = Dir.mktmpdir(TMP_PREFIX)
+    repo = Git.clone(source_url, dir)
```
[the remark this answers is not in the document]
Comment (dlukman) [reply lib/knife/changelog/policyfile.rb]: If you don't mind, I would like to cover this case in another pull request.

File: lib/knife/changelog/policyfile.rb (hunk 1 above)
Comment (dlukman) [reply lib/knife/changelog/policyfile.rb]: Renamed.

File: spec/unit/policyfile_spec.rb (hunk 3 above)
Comment (dlukman) [reply spec/unit/policyfile_spec.rb]: Added the test

File: spec/unit/policyfile_spec.rb (hunk 4 above)
Comment (dlukman) [reply spec/unit/policyfile_spec.rb]: I've covered this method in the next commit. I also added tests for two other methods.

Comment (clems4ever) [reply spec/unit/policyfile_spec.rb]: Great!

**Comment by clems4ever** [comment]
LGTM, thank you for the tests.

**Comment by AntonOfTheWoods** [comment]
LGTM, I'll merge tomorrow morning if @kamaradclimber hasn't done so beforehand! :)

**Comment by clems4ever** [comment]
LGTM
